File: pset1/credit/python/luhn_algo.py
def valid(num):
    # Reversed num
    rev_num = str(num)[::-1]
    # x is first digit of reversed num, y is second, x is third (...)
    x = []
    y = []
    for idx, val in enumerate(rev_num):
        # if x (idx=0/2/4/6 etc), don't do anything
        # if y (idx=1/3/5/7 etc), multiply by 2
        if idx % 2 == 1:
            multiplied = int(val) * 2
            if multiplied >= 10:
                y.append(1 + multiplied - 10)
            else:
                y.append(multiplied)
        else:
            x.append(int(val))
    print(x, y)
    if (sum(x) + sum(y)) % 10 == 0:
        return print(f"{num} is VALID")
    else:
        return print(f"{num} is INVALID")

valid(4003600000000014)

File: pset1/credit/C/luhn_algo.c
int valid(unsigned long long card) {
    // unsigned long long card = 4003600000000014ULL;
    // unsigned long long card = 378282246310005ULL;
    const unsigned long long number = card;
    int counter = 0;
    int x[8] = {};
    int y[8] = {};
    while (card > 0) {
        int digit = card % 10;
        if (counter % 2 == 0) {
            x[counter / 2] = digit;
            counter += 1;
        }
        else {
            int multiplied = digit * 2;
            if (multiplied >= 10) {
                y[counter / 2] = 1 + multiplied - 10;
            }
            else {
                y[counter / 2] = multiplied;
            }
            counter += 1;
        }
    card /= 10;
    }
    // printf("x = ");
    // for (int i = 0; i < 8; i++) {
    //     printf("%d", x[i]);
    //     if (i < 7) {
    //         printf(",");
    //     }
    //     else {
    //         printf("\n");
    //     }
    // }
    // printf("y = ");
    // for (int i = 0; i < 8; i++) {
    //     printf("%d", y[i]);
    //     if (i < 7) {
    //         printf(",");
    //     }
    //     else {
    //         printf("\n");
    //     }
    // }
    int sum = 0;
    for (int i = 0; i < 8; i++) {
        sum = sum + x[i] + y[i];
        // printf("%d\n", sum);
    }
    if (sum % 10 == 0) {
        // printf("%llu is VALID\n", number);
        return 0;
    }
    else {
        // printf("%llu is INVALID\n", number);
        return 1;
    }

    return 1;
}

void name(unsigned long long card, int* add_count, int* add_digits) {
    int count = 0;
    while (card >= 100) {
        card /= 10;
        count += 1;
    }
    count += 2;
    *add_count = count;
    *add_digits = card;
}